### ISSUE TITLE
Digitizing mode is geometry type agnostic

### ISSUE THREAD

**Issue by GCouet** [qgis/QGIS#51014]
### What is the bug or the crash?

With the relatively new digitizing mode, QGIS doesn't adapt to the current geometry type. Whenever the digitizing mode is switched automatically by QGIS by using the Shape Digitizing toolbar, it will stay in that mode regardless of the layers the user will work on later.

This forces users to click an additional button (or a custom shortcut if they have set it up) everytime they have to, for example, switch from digitizing a rectangle on one polygon layer to digitizing a line on a linestring layer.

### Steps to reproduce the issue

![digitize_awkward](https://user-images.githubusercontent.com/52578322/204006696-466e81af-7a56-4216-9e27-03c727a8eef8.gif)

### Versions

QGIS version | 3.28.0-Firenze | QGIS code revision | ed3ad0430f3

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

### Additional context

This really feels like a regression to me, or at least a downgrade in efficiency. I understand the purpose of the refactoring, but it introduced several button clicks for basic digitizing unless modifying the user profile by adding custom shortcuts.

Potential solutions:
- Make the buttons on the Shape Digitizing toolbar toggles. When one is toggled off, automatically switch to digitize with segment
- Automatically switch back to digitize with segment when layer editing is toggled off or on
- Add a **default** one-key keyboard shortcut for the digitizing modes so users don't have to hunt for a button on toolbars or customize their own shortcuts

**Comment by roya0045** [comment]
Already discussed in #49541 might have been fixed by  #50691 , though unsure if a ptach will be made until the next realease, to fill the gap.

**Comment by agiudiceandrea** [comment]
@roya0045, #50691 has been already merged in QGIS 3.28.1. The behaviour described by @GCouet occurs also in QGIS 3.28.1 (although the buttons are in different toolbars) and it is different from the behavior of the digitizing tool before #46687.

**Comment by roya0045** [comment]
> @roya0045, #50691 has been already merged in QGIS 3.28.1. The behaviour described by @GCouet occurs also in QGIS 3.28.1 (although the buttons are in different toolbars) and it is different from the behavior of the digitizing tool before #46687. Maybe @3nids knows if this was an intended behaviour change or an oversight.
> 
> Just to clarify: it seems to me the different behaviour is not about the "geometry agnosticism" of the digitizing tool, but it is about the fact that before #46687 the digitizing tool mode was reset when the edit mode was switched off, while after #46687 the digitizing tool mode was not reset when the edit mode was switched off.

Honestly if this wasn't an oversight this is a horrible choice. Now the question is: should it reset on change of layer, should it change on a change of geometry type or keep a registry matching digitizing type to geometry type and simply reuse the last good setting.

I think that keeping a registry per geometry type(only line and polygon) would be the best option as it doesn't change much, but you don't have unexpected results when a switching to a line layer and generating invalid geometries.

**Comment by agiudiceandrea** [comment]
Gentle ping @3nids.

**Comment by 3nids** [comment]
The digitizing mode is both geometry and layer agnostic in its current implementation.
To make it simple, the idea behind this is that the shapes are available for both line and geometry layers.

If I translate well your proposal, this means defining a preferred mode as segment digitizing for line layers.

I'm just afraid it would break others' workflow:
Let's say you are drawing circles on both line and polygon layers and you are switching from one layer to another.
Then with your approach you'll have to switch back to shape digitising each time.

The only solution I see would be the possibility to define a preferred mode for any layer, so that when switching to this layer it would also switch the mode. While this is feasible, this will bring a bit of UX for setting the preference. I am not sure this is worth it.

Again, the idea was to decouple what you are drawing from the the source you are writing to.

**Comment by GCouet** [comment]
@3nids To be frank, I really wonder what is the most used method of digitizing on different geometry types because I have been doing manual digitizing for years and using closed shapes on line layers is something I honestly have **never** done or even considered. I have never been required to capture shapes on line layers. That's what's puzzling me in your argument.

If we take one example, I often have to digitize building footprints, and I definitely use 3-point rectangles for that. Even for irregular polygons, if they are all orthogonal, I will use snapping and 3p rectangles then merge them later. It's definitely faster. I can understand the mode being locked in, this is actually good.

As soon as I have to switch to a line layer though, I have no clue what the justification for the mode to stay locked into shape. As far as I know for most users, closed shapes are overhwhelmingly going to be captured on polygon layers. It's just begging for a topological mess otherwise.

It really does act like a regression when I'm jumping from geometry type to the other and each time I have to hunt for a button on a toolbar to switch digitizing mode. @roya0045 has a good suggestion, IMO.

**Comment by 3nids** [comment]
I totally get the usability point of view, you're right.
One thing to consider is that shapes are not limited to closed geometries, you also have the arcs in it, and potentially more in the future.
But I am fine with switching from shape to.... segment? the last used mode other than shape? when switching from a poylgon to a line layer.
I will take care of this before next release.

**Comment by GCouet** [comment]
Yeah, that would make quickly switching from one type to the other more natural.

As I said higher up, if there was simple default hotkeys, like there is for stream digitizing, to switch from one mode to the other, it would probably make the changes more fluid.

Thanks for the attention to this issue!

**Comment by roya0045** [comment]
> I totally get the usability point of view, you're right.
> 
> One thing to consider is that shapes are not limited to closed geometries, you also have the arcs in it, and potentially more in the future.
> 
> But I am fine with switching from shape to.... segment? the last used mode other than shape? when switching from a poylgon to a line layer.
> 
> I will take care of this before next release.

Just an option for the implementation: store a ´last mode used' and load/write it when the layer geometry type is not the same and isn't a point ( or is a line or poly)

Should be easier to maintain and would be stored in the widget per instance.

And have segment be the default value.

**Comment by agiudiceandrea** [comment]
What would happen for point layers? I think "segment" mode doesn't make much sense for point layers.

**Comment by roya0045** [comment]
> What would happen for point layers? I think "segment" mode doesn't make much sense for point layers.

The widget is disabled for point layer as there is only one mode. That's why the issue is only present when switching from poly to line and the opposite.

**Comment by agiudiceandrea** [comment]
It seems to me the button is not disabled for point layers and the "Digitize with Segment" mode is active while it is not possible to select other modes.
![image](https://user-images.githubusercontent.com/16253859/209339053-4e776fb1-fa95-48ba-80c3-6ae4675a67eb.png)

When a mode other than "Digitize with Segment" is active, the button is greyed out, but the mode displayed in the button icon and tool tip is not correct

![image](https://user-images.githubusercontent.com/16253859/209339704-42d62839-1fc8-47b3-aefa-7cf9c7af79f0.png)
